[transform.basicprofiles] Debounce (time) profile

J-N-K opened the pull request:
This adds a second strategy for debouncing values. If applied to a link, the state/commands are only forwarded if no furter update is received during the configured timespan.

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/factory/BasicProfilesFactory.java:
```diff
@@ -92,16 +98,18 @@
     private final ProfileTypeI18nLocalizationService profileTypeI18nLocalizationService;
     private final Bundle bundle;
 
+    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool("basicprofile");
```
Do you create an dedicated thread pool on purpose? The `ProfileContext` already provides its own thread pool via `getExecutorService()` method.

```suggestion
    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool("basicprofiles");
```

J-N-K replied:
Didn't see that. Removed own threadpool

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:
```diff
@@ -11,4 +11,17 @@
 	</config-description>
+
+	<config-description uri="profile:basic-profiles:debounce-time">
+		<parameter name="toItemDelay" type="integer" min="0" step="1">
```
You may want to add the unit attribute here:

```suggestion
		<parameter name="toItemDelay" type="integer" min="0" step="1" unit="ms">
```

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:
```diff
@@ -11,4 +11,17 @@
+			<description>Number of changes before updating Item State.</description>
+			<default>0</default>
+		</parameter>
+		<parameter name="toHandlerDelay" type="integer" min="0" step="1">
```
```suggestion
		<parameter name="toHandlerDelay" type="integer" min="0" step="1" unit="ms">
```

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/i18n/basicprofiles.properties:
```diff
@@ -9,8 +9,14 @@ profile.config.basic-profiles.toggle-switch.events.label = Events
 profile-type.basic-profiles.debounce-counting.label = Debounce (Counting)
-profile.config.basic-profiles.debounce-counting.scale.label = Number Of Changes
-profile.config.basic-profiles.debounce-counting.scale.description = Number of changes before updating Item State.
+profile.config.basic-profiles.debounce-counting.numberOfChanges.label = Number Of Changes
```
:+1: Thanks for fixing this.

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/config/DebounceTimeStateProfileConfig.java:
```diff
@@ -0,0 +1,27 @@
+import org.smarthomej.transform.basicprofiles.internal.profiles.DebounceCountingStateProfile;
+
+/**
+ * Configuration for {@link DebounceCountingStateProfile}.
```
```suggestion
 * Configuration for {@link DebounceTimeStateProfile}.
```

cweitkamp reviewed (changes requested):
Looks good so far. I added some remarks.

We now could add a `Throttle` Profile too, which accepts a state/command first and blocks following state changes/commands for a user defined amount of time.

J-N-K commented:
I think I addresses all comments. The "throttle" could be easily integrated by adding a "send first/send last" parameter to the profile, I'll check that later.

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/profiles/DebounceTimeStateProfile.java:
```diff
@@ -0,0 +1,184 @@
+
+        if (config.toHandlerDelay < 0) {
+            throw new IllegalArgumentException(String.format(
+                    "debounceTimeToHandler has to be a non-negative integer but was '%d'.", config.toHandlerDelay));
```
```suggestion
                    "'toHandlerDelay' has to be a non-negative integer but was '%d'.", config.toHandlerDelay));
```

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/profiles/DebounceTimeStateProfile.java:
```diff
@@ -0,0 +1,184 @@
+
+        if (config.toItemDelay < 0) {
+            throw new IllegalArgumentException(String
+                    .format("debounceTimeToItem has to be a non-negative integer but was '%d'.", config.toItemDelay));
```
```suggestion
                    .format("'toItemDelay' has to be a non-negative integer but was '%d'.", config.toItemDelay));
```

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:
```diff
@@ -11,4 +11,26 @@
+	<config-description uri="profile:basic-profiles:debounce-time">
+		<parameter name="toItemDelay" type="integer" min="0" step="1" unit="ms">
+			<label>To Item Delay</label>
+			<description>Number of changes before updating Item State.</description>
```
Can you change it to a meaningful explanation for the parameter?

cweitkamp commented on bundles/org.smarthomej.transform.basicprofiles/src/main/resources/OH-INF/config/debounce.xml:
```diff
@@ -11,4 +11,26 @@
+		</parameter>
+		<parameter name="toHandlerDelay" type="integer" min="0" step="1" unit="ms">
+			<label>To Handler Delay</label>
+			<description>Number of changes before updating Item State.</description>
```
See above.

cweitkamp reviewed (commented):
Looks promising. Thanks. Some final remarks.

cweitkamp reviewed (approved):
LGTM.
